snap/snapcraft.yaml

```diff
@@ -12,256 +13,318 @@ description: |
   - build-on: amd64
 
+# for start-timeout, we need snapd 2.38
+# we need command-chain for the hooks too
+assumes: [snapd2.38, command-chain]
+
+# Ideally snapcraft would generate this command-chain spec for the hooks
+# automatically, but that doesn't currently work, see
+# https://bugs.launchpad.net/snapd/+bug/1824255
+# In the meantime, make sure that the snapcraft-runner gets generated by 
+# the apps by specifying the full adapter in at least one of the apps, and 
+# manually craft the command-chain for the configure hook here to passthrough
+# to the snap.yaml
+# All of this is necessary so we can use jq and other tools from inside the 
+# snap in the configure and install hooks
+passthrough:
+  hooks:
+    configure:
+      command-chain:
+      - snap/command-chain/snapcraft-runner
+    install:
+      command-chain:
+      - snap/command-chain/snapcraft-runner
+
 # TODO: upgrade to stable before releasing to beta/candidate/stable
 grade: devel
 confinement: strict
-#
-# TODO: still a few bugs to work out for confinement;
-# seems that there are apparmor denials being generated
-# for open calls to the directory: /var/lib/snapd/void/.
-# Prior to the changes to the start-edgex script, each
-# service's log files were being written here, so not
-# sure why denials are still being generated.  This
-# be caused by the fact that the current working dir
-# may be defaulting to this dir...
 
 apps:
   # edgex microservices
   consul:
+    adapter: full
     command: bin/start-consul.sh
-    daemon: simple
+    daemon: forking
     plugs: [network, network-bind]
   core-config-seed:
-    command: bin/fork-wrapper.sh $SNAP/bin/go-services-wrapper.sh config-seed -c ${SNAP_DATA}/config
-    plugs: [network, network-bind]
-    daemon: forking
+    adapter: none
+    after: [consul]
+    command: >-
+      bin/config-seed
+      --cmd $SNAP_DATA/config
+      -confdir $SNAP_DATA/config/config-seed/res
+      --props $SNAP_DATA/config/config-seed/res/properties
+    daemon: oneshot
     passthrough:
-      after:
-        - consul
+      start-timeout: 15m
+    plugs: [network, network-bind]
+  redis:
+    adapter: full
+    after: [core-config-seed]
+    command: bin/redis-server $SNAP_DATA/redis/redis.conf
+    daemon: simple
+    plugs: [network, network-bind]
   mongod:
-    command: bin/mongod-wrapper.sh
-    stop-command: bin/mongod --shutdown --dbpath ${SNAP_DATA}/mongo/db
-    plugs: [hardware-observe, network, network-bind, system-observe]
+    adapter: full
+    after: [core-config-seed]
+    command: bin/mongod --dbpath $SNAP_DATA/mongo/db --logpath $SNAP_COMMON/mongodb.log --smallfiles
     daemon: simple
-    passthrough:
-      after:
-        - core-config-seed
+    plugs: [hardware-observe, network, network-bind, system-observe]
+    stop-command: bin/mongod --shutdown --dbpath $SNAP_DATA/mongo/db
   cassandra:
+    adapter: full
     command: bin/cassandra-wrapper.sh
     daemon: forking
+    passthrough:
+      start-timeout: 15m
```

Comment:
15m seems like a *really* long timeout?

Reply:
Well there are a lot of services in the snap and especially on a refresh when all of the services get automatically enabled (before getting disabled when the configure hook runs to work around https://bugs.launchpad.net/snapd/+bug/1818306) it may be 15 minutes before a Raspberry Pi 3B+ running Ubuntu Server arm64 with only 1 GB can successfully start all the services. I don't think it's actually quite that high IIRC it's closer to 5-10 minutes, but I wanted to be safe.